### ISSUE TITLE
Newbie How-To Contribute Tutorials

### ISSUE THREAD

**Issue by IsaiahKelly** [sschmid/Entitas#386]
I really want to contribute to this project, but I'm just starting to learn how to use things like git, Visual Studio and other required tools, so it's a bit overwhelming. Therefore it would be extremely valuable to someone like me if more experienced developers out there could create some absolute beginner text / video tutorials on how to get the Entitas source setup so you can start contributing.

One weakness I see in most tutorials for projects like this is the focus on fellow developers of the same skill level or knowledge. However, what I really need are tutorials that don't assume any previous experience and treat me as completely new and ignorant on the topic. Now I like to think I'm not _completely_ ignorant, but it's always best to assume the worse. 😄

Ideally I'd love a video tutorial that would go through the step-by-step process of setting up Entitas in Visual Studio on Windows and show how to debug your changes and test them in Unity. So essentially a much more detailed step-by-step video version of the [CONTRIBUTING](https://github.com/sschmid/Entitas-CSharp/blob/develop/CONTRIBUTING.md) readme file. The [Cloning Daggerfall Unity from GitHub](https://youtu.be/Okv_NTSg2rw) Tutorial video is a good example of the kind of content I'm talking about here.

For experienced developers here this shouldn't be very difficult to create at all, because I basically just want you to show me how you go about setting up everything when you work with the Entitas source. Which I'm sure is second nature to you. 😉

**Comment by FNGgames** [comment]
These tuts are extremely time consuming to make, and this framework is open source and all. So we've been trying to contribute as a community to the docs and stuff - but we all have day jobs too, so it's a little slow. 

We've added quite a few tuts already, but I assume you're struggling with them from your post. I'd be more than happy to answer questions you have if you PM me on gitter, maybe I can point you in the right direction. I am not an expert or even a professional programmer here, but I've been through the "learning to use entitas" process recently so I might be able to help here and there.

**Comment by T2RKUS** [comment]
Perhaps if you clearly define what you want explained in this video tutorial I or someone else from the entitas community will make one and put it up on youtube. I wouldn't mind doing a video tutorial but wouldn't know what to cover. I've done live streaming on livecoding and twitch of my projects before but not made any youtube tutorials.

**Comment by IsaiahKelly** [comment]
@FNGgames @T2RKUS Thanks for the reply. Here's my specific situation if that helps any:

I already have the Entitas source forked and cloned to my local machine through SourceTree, but I don't know how to build and test any modifications made in Visual Studio from Unity. 

Simon has been trying to help me with this, but he's on Mac and I'm on Windows and so far I can't get it working, At the moment typing `./Scripts/update` into the SourceTree Terminal just returns `/c/GitHub/Entitas/Scripts/build: line 6: xbuild: command not found` and I have no idea how to fix this.

I've been looking at other means of testing in Unity, but so far I don't even know which scripts or DLLs I should copy over from the Entitas souce to Unity in order to test my modifications manually.

**Comment by T2RKUS** [comment]
I see, and what version of Unity are you currently using?

**Comment by T2RKUS** [comment]
For starters, when I ran xbuild from the commandline the command didn't exist so I looked for xbuild in my Program Files(x86) on windows 10 It was in the Mono folder after doing a search for it. If you add xbuild.exe to your PATH then it should make that error go away.

**Comment by T2RKUS** [comment]
Just found out that if you go to Task Manager and find Windows Explorer (The File Browser) and right click -> restart it will also reload the desktop and the environment variables as well. No restart needed.

**Comment by sschmid** [comment]
@IsaiahKelly This might be fixed in the Future. Microsoft just release a new mono version last week where they actually made the switch from xbuild (mono) to msbuild (which also runs on windows). Currently I cannot update to msbuild, but I might in the future.
In general, you have to make sure that the paths in `Scripts/consts` are setup to work properly on windows. E.g. `/c/GitHub/Entitas/Scripts/` doesn't look like a valid path to me
should more look like `c:\GitHub\...`

**Comment by T2RKUS** [comment]
The build script worked for me under the source tree terminal but not cygwin

**Comment by T2RKUS** [comment]
I wonder if that has to do with me adding xbuild to PATH. I didn't try it before that so I don't know.

**Comment by sschmid** [comment]
@T2RKUS on windows? do you have mono installed? (xbuild)

**Comment by T2RKUS** [comment]
I installed Mono yesterday but It wasn't added to my PATH.

**Comment by T2RKUS** [comment]
The build script works as is, no need to change it on windows.

**Comment by FNGgames** [comment]
I will probably not be much help with regards to git I'm afraid, or even the entitas backend - i can help with front-end stuff though

**Comment by IsaiahKelly** [comment]
Maybe part of the problem here is I don't know what xbuild and msbuild are, and probably also don't completely understand what you mean by mono,but I'm guessing your not talking about MonoDevelop? 😳 

So do I need to install this?: http://www.mono-project.com/download/ and should it be the 32-bit or 64-bit version?

It would also seem that constant `UNITY="/Applications/Unity/Unity.app/Contents/MacOS/Unity"` is useless on Windows and should be changed.

**Comment by T2RKUS** [comment]
@IsaiahKelly Now that I have it working on my machine I'll write a build and copy script in bash which uses the build script then copies everything to where it needs to go. basically a unity template demployement script if you like then upload it here.

**Comment by T2RKUS** [comment]
@sschmid you could do a bash -version and change the UNITY variable based on whether the terminal returns "i686-pc-msys", "msys" or something for windows.

**Comment by IsaiahKelly** [comment]
Yeah, I was already thinking there must be a way to make the scripts execute differently on Windows. That would help newbs like myself a lot.

**Comment by T2RKUS** [comment]
hmm, Just wondering whether I should make use of the deploy script and call that after build then unzip that file into the assets folder defined in my build script.

**Comment by T2RKUS** [comment]
```bash
#!/bin/bash -e
echo $0
source "$(cd $(dirname ${BASH_SOURCE[0]}) && pwd)/consts"

# User Configuration

TEMPLATE_ROOT="${ROOT}/EntitasTemplate"

# Script Configuration

ASSETS="${TEMPLATE_ROOT}/Assets"
ASSETS_LIBS="${ASSETS}/Libraries"
ASSETS_ENTITAS_ROOT="${ASSETS_LIBS}/Entitas"
ASSETS_ENTITAS="${ASSETS_ENTITAS_ROOT}/Entitas"
ASSETS_ENTITAS_EDITOR="${ASSETS_ENTITAS}/Editor"
ASSETS_ENTITAS_MIGRATION="${ASSETS_ENTITAS_ROOT}/Migration"
ASSETS_ENTITAS_EDITOR_IMAGES="${ASSETS_ENTITAS_EDITOR}/Images"

${SCRIPTS}/update

CP="cp -f"

# Copy Libraries  ---------------------------------------------------------

mkdir $ASSETS_LIBS
mkdir $ASSETS_ENTITAS_ROOT $ASSETS_ENTITAS $ASSETS_ENTITAS_EDITOR $ASSETS_ENTITAS_EDITOR_IMAGES
mkdir $ASSETS_ENTITAS_MIGRATION

# Entitas 

CP "${ROOT}/${ES}/${BIN}/${ES}.dll" $ASSETS_ENTITAS
CP "${ROOT}/${ES_UTILS}/${BIN}/${ES_UTILS}.dll" $ASSETS_ENTITAS
CP "${ADDONS}/${ES_U}/${BIN}/${ES_U}.dll" $ASSETS_ENTITAS
CP "${ADDONS}/${ES_UE}/${BIN}/${ES_UE}.dll" $ASSETS_ENTITAS_EDITOR

# Entitas Code Generation

CP "${ADDONS}/${ES_CG}/${BIN}/${ES_CG}.dll" $ASSETS_ENTITAS_EDITOR
CP "${ADDONS}/${ES_CG_ATTR}/${BIN}/${ES_CG_ATTR}.dll" $ASSETS_ENTITAS
CP "${ADDONS}/${ES_CG_G}/${BIN}/${ES_CG_G}.dll" $ASSETS_ENTITAS_EDITOR
CP "${ADDONS}/${ES_CG_P}/${BIN}/${ES_CG_P}.dll" $ASSETS_ENTITAS_EDITOR
CP "${ADDONS}/${ES_CG_UE}/${BIN}/${ES_CG_UE}.dll" $ASSETS_ENTITAS_EDITOR

# Entitas Migration

CP "${ADDONS}/${ES_MIG}/${BIN}/${ES_MIG}.dll" $ASSETS_ENTITAS_MIGRATION
CP "${ADDONS}/${ES_MIG}/${BIN}/${ES_MIG}.dll" $ASSETS_ENTITAS_EDITOR
CP "${ROOT}/${ES_UTILS}/${BIN}/${ES_UTILS}.dll" $ASSETS_ENTITAS_MIGRATION
CP "${ADDONS}/${ES_MIG_CLI}/${BIN}/${ES_MIG_CLI}.exe" $ASSETS_ENTITAS_MIGRATION
CP "${ADDONS}/${ES_MIG_UE}/${BIN}/${ES_MIG_UE}.dll" $ASSETS_ENTITAS_EDITOR

# Entitas Visual Debugging

CP "${ADDONS}/${ES_VD_CG_P}/${BIN}/${ES_VD_CG_P}.dll" $ASSETS_ENTITAS_EDITOR
CP "${ADDONS}/${ES_VD_U}/${BIN}/${ES_VD_U}.dll" $ASSETS_ENTITAS
CP "${ADDONS}/${ES_VD_UE}/${BIN}/${ES_VD_UE}.dll" $ASSETS_ENTITAS_EDITOR

# Entitas Blueprints

CP "${ADDONS}/${ES_BP}/${BIN}/${ES_BP}.dll" $ASSETS_ENTITAS
CP "${ADDONS}/${ES_BP_CG_P}/${BIN}/${ES_BP_CG_P}.dll" $ASSETS_ENTITAS_EDITOR
CP "${ADDONS}/${ES_BP_U}/${BIN}/${ES_BP_U}.dll" $ASSETS_ENTITAS
CP "${ADDONS}/${ES_BP_UE}/${BIN}/${ES_BP_UE}.dll" $ASSETS_ENTITAS_EDITOR

# Fabl

CP "${ADDONS}/${ES_CG_G}/${BIN}/fabl.dll" $ASSETS_ENTITAS

# Copy Images  ---------------------------------------------------------

cp "${ADDONS}/${ES_UE}/${ES_UE}/Editor/Images/"* $ASSETS_ENTITAS_EDITOR_IMAGES
cp "${ADDONS}/${ES_MIG_UE}/${ES_MIG_UE}/Editor/Images/"* $ASSETS_ENTITAS_EDITOR_IMAGES
cp "${ADDONS}/${ES_VD_UE}/${ES_VD_UE}/Editor/Images/"* $ASSETS_ENTITAS_EDITOR_IMAGES
```
Try putting this script in the Scripts folder of the repo and change the user config to where your project is.

**Comment by IsaiahKelly** [comment]
@T2RKUS Thanks for all your help on this! I installed Mono 64-bit (Preview, no GTK#) and added this PATH:
![xbuild_path](https://cloud.githubusercontent.com/assets/7953130/25246789/bb2e2872-25d6-11e7-9977-ee8ad7b87abf.png)

But get the same `line 6: xbuild: command not found` error again. Is this the wrong version of mono or am I setting the PATH incorrectly?

**Comment by T2RKUS** [comment]
did restart your computer or Windows Explorer after you set the variable.

**Comment by IsaiahKelly** [comment]
Duh!

**Comment by T2RKUS** [comment]
It should look something like this: http://imgur.com/a/2F8Cy

**Comment by IsaiahKelly** [comment]
Did the Task Manger restart trick, but that didn't work. So I restarted and that didn't help either. 😞 

I've also checked the `Scripts/consts` definitions as Simon suggested, but `UNITY` is the only incorrect path I can see.  Can you tell/show me how you have that file setup? Here's mine at the moment:

```
#!/bin/bash -e
SCRIPTS="$(cd $(dirname ${BASH_SOURCE[0]}) && pwd)"
ROOT="${SCRIPTS}/.."

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
#UNITY="/Applications/Unity/Unity.app/Contents/MacOS/Unity"
UNITY="/Program Files/Unity/Editor/Unity"
PROJECT_NAME="Entitas"

ADDONS="${ROOT}/Addons"
ADDONS_PLAYGROUND="${ROOT}/Tests/Unity"

ES="Entitas"
ES_UTILS="Entitas.Utils"

ES_U="Entitas.Unity"
ES_UE="Entitas.Unity.Editor"

ES_CG="Entitas.CodeGeneration"
ES_CG_ATTR="Entitas.CodeGeneration.Attributes"
ES_CG_G="Entitas.CodeGeneration.CodeGenerator"
ES_CG_G_CLI="Entitas.CodeGeneration.CodeGenerator.CLI"
ES_CG_P="Entitas.CodeGeneration.Plugins"
ES_CG_UE="Entitas.CodeGeneration.Unity.Editor"

ES_MIG="Entitas.Migration"
ES_MIG_CLI="Entitas.Migration.CLI"
ES_MIG_UE="Entitas.Migration.Unity.Editor"

ES_VD_CG_P="Entitas.VisualDebugging.CodeGeneration.Plugins"
ES_VD_U="Entitas.VisualDebugging.Unity"
ES_VD_UE="Entitas.VisualDebugging.Unity.Editor"

ES_BP="Entitas.Blueprints"
ES_BP_CG_P="Entitas.Blueprints.CodeGeneration.Plugins"
ES_BP_U="Entitas.Blueprints.Unity"
ES_BP_UE="Entitas.Blueprints.Unity.Editor"

BIN="bin/Release"
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

SOLUTION="${ROOT}/${PROJECT_NAME}.sln"
PROJECT_VERSION=`cat ${ROOT}/${PROJECT_NAME}/${PROJECT_NAME}/version`

TEST_PROJECT="${ROOT}/Tests/Tests/Tests.csproj"
TEST_RUNNER="${ROOT}/Tests/Tests/bin/Release/Tests.exe"

# build
BUILD="${ROOT}/Build"
BUILD_SRC="${BUILD}/src"
BUILD_DOCS="${BUILD}/docs"
BUILD_DEPLOY="${BUILD}/deploy"

DOCS="$ROOT/docs"
DOCS_RES="${SCRIPTS}/docs_resources"
DOCSET="com.sschmid.${PROJECT_NAME}.docset"
DOCSET_KEY="entitas"

cpe() {
  rsync -ar --include-from "${SCRIPTS}/rsync_include.txt" --exclude-from "${SCRIPTS}/rsync_exclude.txt" $1 $2
}
```

**Comment by IsaiahKelly** [comment]
Should I try the 32-bit version of Mono instead?

**Comment by T2RKUS** [comment]
My script just calls the build script in the scripts directory so you should have it alongside 
update
build
pack etc.

it loads the consts file, builds, then copies the dll/images to the folder specified in my script. you can create a new file called update_unity in that scripts folder and run it from there.

**Comment by IsaiahKelly** [comment]
@T2RKUS I already placed your script there and named it "temp", but have now changed it to "update_unity" 

Installed mono 32-bit and and changed PATH to that installation and removed "xbuild.exe" since your screenshot doesn't show that in the path. Still doesn't work.

```
$ ./Scripts/update_unity
./Scripts/update_unity
/c/GitHub/Entitas/Scripts/update
/c/GitHub/Entitas/Scripts/build
/c/GitHub/Entitas/Scripts/build: line 6: xbuild: command not found
```
I'm wondering if my issue here is with the git working directory path? Simon doesn't think ` /c/GitHub/Entitas` is a valid path, but I'm not sure if this is true or just the way SourceTree handles it? The literal path is `C:\GitHub\Entitas\` Is this path correct and how would I change it if it isn't?

**Comment by T2RKUS** [comment]
I was just using the MSYS shell (which is what you are using if you start the terminal from Source Tree as you say) from my downloads folder and it worked fine there. I don't see what the problem could be. It was actually pretty easy to get working on my windows machine with minimum effort.

**Comment by sschmid** [comment]
Btw, just committed "Using msbuild instead of xbuild"
6df2aa55b2e1a5d2517b4fd4c86e75c4d551d7b4

**Comment by sschmid** [comment]
sorry, have to revert since travis ci doesn't use mono 5.0 yet

**Comment by IsaiahKelly** [comment]
Hey! It's me again 🙂. I recently reinstalled Windows 10, Unity 2017.1 and Visual Studio 2017 and decided to give this another shot. So I downloaded the Entitas source and opened it up again. Tried running Tests again, but I believe it gave me an error related to the build target or something, so I changed the target to x86 and it finally ran! However, now I'm getting an error message on line 17 of check_namespaces.cs with the following message:
```
NUnit.Framework.AssertionException: '  Expected: greater than 80

  But was:  0

```
No idea what this actual means, so I totally stuck again. 😞
